pkg/router/controller/route_secret_manager.go

```diff
@@ -118,106 +192,199 @@ func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *route
+			log.V(4).Info(msg)
 
-			// read referenced secret (updated data)
-			secret, err := p.secretManager.GetSecret(context.TODO(), route.Namespace, route.Name)
+			// Ensure fetching the updated route
+			route, err := p.getUpdatedRoute(namespace, routeName)
 			if err != nil {
-				log.Error(err, "failed to get referenced secret")
-				p.recorder.RecordRouteRejection(route, "ExternalCertificateGetFailed", err.Error())
-				p.plugin.HandleRoute(watch.Deleted, route)
 				return
 			}
 
-			// update tls.Certificate and tls.Key
-			route.Spec.TLS.Certificate = string(secret.Data["tls.crt"])
-			route.Spec.TLS.Key = string(secret.Data["tls.key"])
-
-			// call the next plugin with watch.Modified
-			p.plugin.HandleRoute(watch.Modified, route)
-			// commit the changes
-			p.plugin.Commit()
+			// Update the route status, keeping `Admitted=True`, to signal route update
+			// and re-evaluatation by all the plugins (including this plugin).
+			p.recorder.RecordRouteUpdate(route, "ExternalCertificateSecretUpdated", msg)
 		},
 
 		DeleteFunc: func(obj interface{}) {
 			secret := obj.(*kapi.Secret)
-			msg := fmt.Sprintf("secret %s deleted for route %s/%s", secret.Name, route.Namespace, route.Name)
+			key := generateKey(namespace, routeName)
+			msg := fmt.Sprintf("secret %q deleted for route %q", secret.Name, key)
 			log.V(4).Info(msg)
 
-			// unregister associated secret monitor
-			if err := p.secretManager.UnregisterRoute(route.Namespace, route.Name); err != nil {
-				log.Error(err, "failed to unregister route")
+			// keep the secret monitor active and mark the secret as deleted for this route.
+			p.deletedSecrets.Store(key, true)
+
+			// Ensure fetching the updated route
+			route, err := p.getUpdatedRoute(namespace, routeName)
+			if err != nil {
+				return
 			}
 
+			// Reject this route
 			p.recorder.RecordRouteRejection(route, "ExternalCertificateSecretDeleted", msg)
+			// Stop serving this route
 			p.plugin.HandleRoute(watch.Deleted, route)
 		},
 	}
 }
 
+// validate checks the route's external certificate configuration.
+// If the validation fails, it records the route rejection and triggers
+// the deletion of the route by calling the HandleRoute method with a watch.Deleted event.
+//
+// NOTE: TLS data validation and sanitization are handled by the next plugin `ExtendedValidator`,
+// by reading the "tls.crt" and "tls.key" added by populateRouteTLSFromSecret.
+func (p *RouteSecretManager) validate(route *routev1.Route) error {
+	fldPath := field.NewPath("spec").Child("tls").Child("externalCertificate")
+	if err := routeapihelpers.ValidateTLSExternalCertificate(route, fldPath, p.sarClient, p.secretsGetter).ToAggregate(); err != nil {
+		log.Error(err, "skipping route due to invalid externalCertificate configuration", "namespace", route.Namespace, "route", route.Name)
+		p.recorder.RecordRouteRejection(route, "ExternalCertificateValidationFailed", err.Error())
+		p.plugin.HandleRoute(watch.Deleted, route)
+		return err
```

Comment:
Can this be moved up the callstack in `HandleRoute` function where `validate` is called? This way we will have the plugin chain propagation (`p.plugin.HandleRoute`)  and logging in  the same function and not hidden in the callstack.

```suggestion
		return fmt.Errorf("invalid external certificate configuration: %w", err)
```

Reply:
This function is called in `HandleRoute` and also inside `validateAndRegister` function. Moving the logic from here and using them directly will reduce code readability. Moreover, the godoc added for this function is self explanatory that if validation fails, it records the route rejection and calls the next plugin. I would like to have them inside a function to reduce code redundancy.